ext/opentelemetry-ext-opentracing-shim/src/opentelemetry/ext/opentracing_shim/__init__.py

```diff
@@ -0,0 +1,289 @@
+    `span` argument, the `ScopeWrapper` is initialized as a regular `Scope`
+    object.
+
+    When calling the initializer while passing a context manager *generator* in
+    the `span_cm` argument (as returned by the `use_span()` method of
+    OpenTelemetry `Tracer` objects,  the resulting `ScopeWrapper` becomes
+    usable as a context manager (using `with` statements).
+
+    It is necessary to have both ways for constructing `ScopeWrapper`
+    objects because in some cases we need to create the object from a context
+    manager, in which case our only way of retrieving a `Span` object is by
+    calling the `__enter__()` method on the context manager, which makes the
+    span active in the OpenTelemetry tracer; whereas in other cases we need to
+    accept a `SpanWrapper` object and wrap it in a `ScopeWrapper`.
+    """
+
+    def __init__(self, manager, span=None, span_cm=None):
+        super().__init__(manager, span)
+        self._span_cm = span_cm
+
+    def close(self):
+        self._span.unwrap().end()
+        # TODO: Set active span on OpenTelemetry tracer.
+        # https://github.com/open-telemetry/opentelemetry-python/issues/161#issuecomment-534136274
+
+    def __enter__(self):
+        if self._span_cm is not None:
+            otel_span = self._span_cm.__enter__()
+            self._span = SpanWrapper(
+                self._manager.tracer, otel_span.get_context(), otel_span
+            )
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        # Span._on_error(self.span, exc_type, exc_val, exc_tb)
+        # TODO: Do we want this ^ ? It exists in the overridden method on the
+        # base class in the OpenTracing API.
+        if self._span_cm is not None:
+            self._span_cm.__exit__(exc_type, exc_val, exc_tb)
+
+
+class ScopeManagerWrapper(opentracing.ScopeManager):
+    def __init__(self, tracer):
+        # The only thing the `__init__()` method on the base class does is
+        # initialize `self._noop_span` and `self._noop_scope` with no-op
+        # objects. Therefore, it doesn't seem useful to call it.
+        # pylint: disable=super-init-not-called
+        self._tracer = tracer
+
+    def activate(self, span, finish_on_close):
+        span_cm = self._tracer.unwrap().use_span(
+            span.unwrap(), end_on_exit=finish_on_close
+        )
+        return ScopeWrapper(self, span_cm=span_cm)
+
+    @property
+    def active(self):
+        span = self._tracer.unwrap().get_current_span()
+        if span is None:
+            return None
+
+        wrapped_span = SpanWrapper(self._tracer, span.get_context(), span)
+        return ScopeWrapper(self, span=wrapped_span)
+        # TODO: Return a saved instance of SpanWrapper instead of constructing
+        # a new object (and the same for ScopeWrapper?).
+        # https://github.com/open-telemetry/opentelemetry-python/issues/161#issuecomment-534136274
+
+    @property
+    def tracer(self):
```

Comment:
Do you really need this property?

Reply:
@carlosalberto currently I need it because of the following:

https://github.com/kinvolk/opentelemetry-python/blob/e35a5fa5594507969ca20da3c594d44b81e1f546/ext/opentelemetry-ext-opentracing-shim/src/opentelemetry/ext/opentracing_shim/__init__.py#L132

I'm building a `SpanShim` when constructing a `ScopeShim`, and the initializer needs to populate the `tracer` argument which is required by the OpenTracing API.

There is a good chance this can be refactored, however in the current state of the implementation it seems it is necessary. Do you see an alternative?

Reply:
Sounds fine to me.